### PR TITLE
WIP:Dockerize Concord-BFT build

### PR DESCRIPTION
**Pull request by f-squirrel**
This PR brings the possibility to run Concord-BFT builds and tests in a docker container.

Rationale:

Separate the building/installation of 3rd parties from building/testing the Concord-BFT.
Have an ability to use the same environment to run builds/ tests at dev stations and Travis CI.
Run Apollo tests in an isolated environment.
Have the same compilation keys while building Concord-BFT standalone and as part of Concord.

How to use it on a dev station:

There are two ways to build Concord-BFT: docker build and native build.

Native build:
```
cd concord-bft
sudo ./install_deps.sh # Installs all the required 3rd parties, compilers and tools system-wide
mkdir build  && cd build
cmake <compilation commands > .. && make -j8
```

Docker build:
```
cd concord-bft
make run-c # Run the build container in the background
make build-s # Builds source code
make test # Run test
```
Note that the `Makefile` is configurable.
For example, if you want to use another compiler you may pass it to the `make`:
```
make CONCORD_BFT_CONTAINER_CXX=g++ \
    CONCORD_BFT_CONTAINER_CC=gcc \
    build-s
```
Run `make help` to see more commands.

### CODE REVIEW

**Comment by teoparvanov** [comment]
Thanks Dima, I appreciate your effort here. But let's bring some more clarity on the following questions:
1) This PR allows us to build the entire concord-bft codebase within a _single_ Docker container, right?
2) Is the CI build going to change, to build this container and run the tests inside it? Aren't we supposed to run both the container AND the native code building in CI?
3) If 2. is true, how would the Apollo tests for network partitioning run inside this container?

Thanks,
Teo

**Comment by f-squirrel** [comment]
@teoparvanov ,
Q: This PR allows us to build the entire concord-bft codebase within a single Docker container, right?
A: Yes, one **single** container
Q:Is the CI build going to change, to build this container and run the tests inside it? Aren't we supposed to run both the containerized AND the "native" Ubuntu build in CI?
A: CI is not supposed to build the container but to use the container to build and test Concord-BFT. Since the container is built on top of Ubuntu 18.04 I see no reason to run the same in Travis CI VM "natively".
Q: What is the point of maintaining the USE_CONAN option at all, if we intend to switch to containerized build?
A: No point, I just wanted to split it into a few PRs.
Q: Can we drop Conan entirely?
A: Yes

**Comment by teoparvanov** [comment]
> 
> 
> @teoparvanov ,
> Q: This PR allows us to build the entire concord-bft codebase within a single Docker container, right?
> A: Yes, one **single** container
> Q:Is the CI build going to change, to build this container and run the tests inside it? Aren't we supposed to run both the containerized AND the "native" Ubuntu build in CI?
> A: CI is not supposed to build the container but to use the container to build and test Concord-BFT. Since the container is built on top of Ubuntu 18.04 I see no reason to run the same in Travis CI VM "natively".
> Q: What is the point of maintaining the USE_CONAN option at all, if we intend to switch to containerized build?
> A: No point, I just wanted to split it into a few PRs.
> Q: Can we drop Conan entirely?
> A: Yes

Thanks Dima - my concerns are addressed. I also like the fact this is the first step to removing Conan altogether.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+build-s: ## Build Concord-BFT source. Note: this command is mostly for developers
+	docker exec -it --workdir=${CONCORD_BFT_TARGET_SOURCE_PATH} ${CONCORD_BFT_DOCKER_IMAGE} \
+		${CONCORD_BFT_CONTAINER_SHELL} -c \
+		"mkdir -p ${CONCORD_BFT_TARGET_SOURCE_PATH}/${CONCORD_BFT_BUILD_DIR} && \
+		cd ${CONCORD_BFT_BUILD_DIR} && \
+		CC=${CONCORD_BFT_CONTAINER_CC} CXX=${CONCORD_BFT_CONTAINER_CXX} \
+		cmake ${CONCORD_BFT_CMAKE_FLAGS} .. && \
+		make format-check && \
+		make -j8"
```
[the remark this answers is not in the document]
Comment (f-squirrel) [reply Makefile:70]: For some reason, `$(nproc) returns an empty string in docker.
Need to figure out how to change it.

File: storage/include/s3/client.hpp [hunk 2]
```diff
@@ -17,6 +17,7 @@
 #include <cstring>
 #include <mutex>
 #include <thread>
+#include <mutex>
```
[the remark this answers is not in the document]
Comment (f-squirrel) [reply storage/include/s3/client.hpp:20]: This one is already committed(https://github.com/vmware/concord-bft/blob/master/storage/include/s3/client.hpp#L18).
I don't know why it is shown in the PR.

File: install_deps.sh [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                --enable-static && \
+    make -j8 && \
+    make install && \
+    cd ${HOME} && \
+    rm -r log4cplus-1.2.1
+
+cd ${HOME}
+git clone https://github.com/google/googletest.git && \
+    cd googletest && \
+    git checkout e93da23920e5b6887d6a6a291c3a59f83f5b579e && \
```
[the remark this answers is not in the document]
Comment (f-squirrel) [reply install_deps.sh:71]: Agreed, but not for this PR.

File: install_deps.sh [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    mkdir _build && cd _build && \
+    cmake -DCMAKE_CXX_FLAGS="-std=c++11 -march=x86-64 -mtune=generic" .. && \
+    make -j8 && \
```
[the remark this answers is not in the document]
Comment (f-squirrel) [reply install_deps.sh:74]: Done

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+CONCORD_BFT_CONTAINER_MOUNT_CONSISTENCY=,consistency=cached
+CONCORD_BFT_CTEST_TIMEOUT:=3000 # Default value is 1500 sec. It takes 2500 to run all the tests at my dev station
+
+.PHONY: help
+
+help: ## The Makefile helps to build Concord-BFT in a docker container
+	@cat $(MAKEFILE_LIST) | grep -E '^[a-zA-Z_-]+:.*?## .*$$' | \
+		awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
+
+	# Basic HOW-TO:
```
[the remark this answers is not in the document]
Comment (f-squirrel) [reply Makefile:32]: Done

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	# make pull     # Pull image from Docker Hub
+	# make build-s  # Build Concord-BFT sources
+	# make test     # Run tests
+	# make remove-c # Remove existing container
+	# make build-nc # Build docker image locally
+
+build-nc: ## Re-build the container without caching
```
[the remark this answers is not in the document]
Comment (f-squirrel) [reply Makefile]: Done

File: install_deps.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (f-squirrel) [reply install_deps.sh]: Done

**Comment by f-squirrel** [comment]
Closed because the functionality was submitted as part of GitHub Actions PR

**Pull request closed without merging**